### PR TITLE
feat(snaps): Add `useDisplayName` hook

### PR DESCRIPTION
**Pull request by GuillaumeRx**
## **Description**

This PR adds a hook to get the display name or the address book entry for an address.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27868?quickstart=1)

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by david0xd — DISMISSED** [review]

File: ui/selectors/selectors.js
Comment (FrederikBolding) [inline ui/selectors/selectors.js:1605]: Ideally this selector just supports CAIP-10 addresses and we can handle this edge-case elsewhere, e.g. in `SnapUIAddress`. Most components should require CAIP-10 and not have backwards compat

File: ui/selectors/selectors.js
Comment (FrederikBolding) [inline ui/selectors/selectors.js:1626]:
```suggestion
  return accountName ?? addressBookEntry?.name ?? null;
```
I think the consuming component should decide how to show no name, instead of us forcing a short address

File: ui/selectors/selectors.js
Comment (FrederikBolding) [inline ui/selectors/selectors.js:1600]: Can we memoize so we don't cause a bunch of re-renders by using this selector?

File: ui/selectors/selectors.js
Comment (GuillaumeRx) [inline ui/selectors/selectors.js:1627]: https://github.com/MetaMask/metamask-extension/blob/04b17ffb1cd400588b3e9814de32794f9405fb94/ui/selectors/selectors.js#L641 😅

File: ui/selectors/selectors.js
Comment (FrederikBolding) [inline ui/selectors/selectors.js:1600]: I don' think the memoization will be effective if we pass in the entire raw state. Wondering how we can structure this to only reference the state we need 🤔

File: ui/hooks/snaps/useDisplayName.ts
Comment (FrederikBolding) [inline ui/hooks/snaps/useDisplayName.ts:44]: This breaks the rules of React hooks, you have to call the selector even if you aren't using the return value

File: ui/selectors/selectors.js
Comment (FrederikBolding) [inline ui/selectors/selectors.js:613]: Couple of these that should be removed before merging 😄

File: ui/selectors/selectors.js
Comment (FrederikBolding) [inline ui/selectors/selectors.js:609]: Can this be a deep equal selector?

File: ui/selectors/selectors.js
Comment (FrederikBolding) [inline ui/selectors/selectors.js:626]: Can this be a deep equal selector?

File: ui/selectors/selectors.js
Comment (GuillaumeRx) [reply ui/selectors/selectors.js:609]: It is, look under

File: ui/selectors/selectors.js
Comment (GuillaumeRx) [reply ui/selectors/selectors.js:626]: It is ! look under

File: ui/hooks/snaps/useDisplayName.ts
Comment (FrederikBolding) [inline ui/hooks/snaps/useDisplayName.ts:38]: Since we are adding new selectors can we add them with types so we don't have to do this?

**Review by FrederikBolding — APPROVED** [review]

**Review by Mrtenz — APPROVED** [review]

**Pull request merged**
